modules/consensus/diffs_test.go

```diff
@@ -0,0 +1,736 @@
+	}
+	cst.cs.commitSiafundPoolDiff(sfpd3, modules.DiffApply)
+	if cst.cs.siafundPool.Cmp(adjusted3) != 0 {
+		t.Error("second siafund pool adjustment was flawed")
+	}
+	cst.cs.commitSiafundPoolDiff(sfpd3, modules.DiffRevert)
+	if cst.cs.siafundPool.Cmp(adjusted2) != 0 {
+		t.Error("reverting second adjustment was flawed")
+	}
+	cst.cs.commitSiafundPoolDiff(sfpd2, modules.DiffRevert)
+	if cst.cs.siafundPool.Cmp(adjusted1) != 0 {
+		t.Error("reverting second adjustment was flawed")
+	}
+	cst.cs.commitSiafundPoolDiff(sfpd1, modules.DiffRevert)
+	if cst.cs.siafundPool.Cmp(initial) != 0 {
+		t.Error("reverting first adjustment was flawed")
+	}
+
+	// Do a chaining set of panics. First apply a negative pool adjustment,
+	// then revert the pool diffs in the wrong order, than apply the pool diffs
+	// in the wrong order.
+	defer func() {
+		r := recover()
+		if r != errApplySiafundPoolDiffMismatch {
+			t.Error("expecting errApplySiafundPoolDiffMismatch, got", r)
+		}
+	}()
+	defer func() {
+		r := recover()
+		if r != errRevertSiafundPoolDiffMismatch {
+			t.Error("expecting errRevertSiafundPoolDiffMismatch, got", r)
+		}
+		cst.cs.commitSiafundPoolDiff(sfpd1, modules.DiffApply)
+	}()
+	defer func() {
+		r := recover()
+		if r != errNonApplySiafundPoolDiff {
+			t.Error(r)
+		}
+		cst.cs.commitSiafundPoolDiff(sfpd1, modules.DiffRevert)
+	}()
+	defer func() {
+		r := recover()
+		if r != errNegativePoolAdjustment {
+			t.Error("expecting errNegativePoolAdjustment, got", r)
+		}
+		sfpd2.Direction = modules.DiffRevert
+		cst.cs.commitSiafundPoolDiff(sfpd2, modules.DiffApply)
+	}()
+	cst.cs.commitSiafundPoolDiff(sfpd1, modules.DiffApply)
+	cst.cs.commitSiafundPoolDiff(sfpd2, modules.DiffApply)
+	negativeAdjustment := adjusted2.Sub(types.NewCurrency64(100))
+	negativeSfpd := modules.SiafundPoolDiff{
+		Previous: adjusted3,
+		Adjusted: negativeAdjustment,
+	}
+	cst.cs.commitSiafundPoolDiff(negativeSfpd, modules.DiffApply)
+}
+
+// TestCommitDiffSetSanity triggers all of the panics in the
+// commitDiffSetSanity method of the consensus set.
+func TestCommitDiffSetSanity(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	cst, err := createConsensusSetTester("TestCommitDiffSetSanity")
+	if err != nil {
+		t.Fatal(err)
+	}
+	bn := cst.cs.currentBlockNode()
+
+	defer func() {
+		r := recover()
+		if r != errDiffsNotGenerated {
+			t.Error("expected errDiffsNotGenerated, got", r)
+		}
+	}()
+	defer func() {
+		r := recover()
+		if r != errWrongAppliedDiffSet {
+			t.Error("expected errWrongAppliedDiffSet, got", r)
+		}
+
+		// Trigger a panic about diffs not being generated.
+		bn.diffsGenerated = false
+		cst.cs.commitDiffSetSanity(bn, modules.DiffRevert)
+	}()
+	defer func() {
+		r := recover()
+		if r != errWrongRevertDiffSet {
+			t.Error("expected errWrongRevertDiffSet, got", r)
+		}
+
+		// trigger a panic about applying the wrong block.
+		bn.block.ParentID[0]++
+		cst.cs.commitDiffSetSanity(bn, modules.DiffApply)
+	}()
```

Comment:
I don't think you can defer multiple `recover`s. Only the first panic will be caught.